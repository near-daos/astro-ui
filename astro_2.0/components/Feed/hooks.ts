import { useAsyncFn, useList, useLocalStorage, useLocation } from 'react-use';
import { useCallback } from 'react';
import { VoteAction } from 'types/proposal';
import { useWalletContext } from 'context/WalletContext';
import { useRouter } from 'next/router';
import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';
import { VOTE_ACTION_SOURCE_PAGE } from 'constants/votingConstants';

export function useMultiVoteActions(): {
  loading: boolean;
  handleVote: (vote: VoteAction) => Promise<void>;
  handleSelect: (id: string) => void;
  handleDismiss: () => void;
  list: string[];
} {
  const { nearService } = useWalletContext();
  const router = useRouter();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();
  const [list, { push, removeAt, clear }] = useList<string>([]);

  const handleSelect = useCallback(
    id => {
      const itemIndex = list.findIndex(item => item === id);

      if (itemIndex !== -1) {
        removeAt(itemIndex);
      } else {
        push(id);
      }
    },
    [list, push, removeAt]
  );

  const handleDismiss = useCallback(() => {
    clear();
  }, [clear]);

  const [{ loading }, handleVote] = useAsyncFn(
    async (vote: VoteAction) => {
      try {
        const paramsArr = list.map(item => {
          const separatorIndex = item.lastIndexOf('-');
          const daoId = item.substring(0, separatorIndex);
          const proposalId = item.substring(separatorIndex + 1);

          return { daoId, proposalId };
        });

        setVoteActionSource(pathname);

        await nearService?.multiVote(vote, paramsArr);

        await router.reload();
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          description: e.message,
          lifetime: 20000,
        });

        await router.reload();
      }
    },
    [router, nearService, list]
  );

  return {
    loading,
    list,
    handleVote,
    handleSelect,
    handleDismiss,
  };
}
