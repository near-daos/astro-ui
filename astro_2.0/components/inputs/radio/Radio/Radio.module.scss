@import 'styles/mixins';

.root {
  --color-checkmark: var(--color-neutral-40);
  --color-checkmark-active: var(--color-primary-50);
  --color-disabled: var(--color-neutral-60);
  --color-label: var(--color-main-gray);
  align-items: center;
  cursor: pointer;

  display: flex;
  position: relative;
  user-select: none;
}

/* Hide the browser's default Checkbox */
.input {
  cursor: pointer;
  height: 0;
  opacity: 0;
  position: absolute;
  width: 0;
}

/* Create a custom Checkbox */
.checkmark {
  border: 1px solid var(--color-checkmark);
  border-radius: 50%;
  height: 16px;
  min-width: 16px;
  position: relative;
  width: 16px;
}

/* When the Checkbox is checked, add a blue background */
.checked ~ .checkmark {
  --color-checkmark: var(--color-checkmark-active);
  border-width: 4px;
}

/* Show the checkmark when checked */
.checked ~ .checkmark::after {
  display: block;
}

/* Create the checkmark/indicator (hidden when not checked) */
.checkmark::after {
  content: '';
  display: none;
  position: absolute;
}

/* Style the checkmark/indicator */
.checkmark::after {
  background-color: #fff;
  border-radius: 50%;
  height: 8px;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
  width: 8px;
}

.label {
  color: var(--color-label);
  display: inline-flex;
  font-family: var(--font-family-secondary);
  font-weight: 600;
  margin-left: 8px;
}

.disabled .checkmark {
  --color-checkmark: var(--color-disabled);
}

.notifications {
  margin-top: 12px;

  @include tablet-portrait {
    margin-top: 0;
  }

  .label {
    @include font(14px, 24px, 500);
    color: var(--color-neutral-60);
    font-family: var(--font-family-primary);
    margin-left: 4px;
    transition: color 0.2s;
  }
}

.notifications:hover {
  .label {
    color: var(--color-brand-black);
  }
}
