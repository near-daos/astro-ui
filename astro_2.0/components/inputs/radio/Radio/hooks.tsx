/* eslint-disable import/prefer-default-export */
import { useContext } from 'react';

import {
  RadioContext,
  RadioContextType,
} from 'astro_2.0/components/inputs/radio/RadioGroup';

export const useRadioContext = (): RadioContextType => {
  const context = useContext(RadioContext);

  if (!context) {
    throw new Error(
      'Radio compound components cannot be rendered outside the Radio component'
    );
  }

  return context;
};
