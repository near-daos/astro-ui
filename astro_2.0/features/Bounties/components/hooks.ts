import { useCallback, useRef, useState } from 'react';
import { useRouter } from 'next/router';
import { useAsyncFn, useLocalStorage, useLocation } from 'react-use';
import axios, { CancelTokenSource } from 'axios';

import { SputnikHttpService } from 'services/sputnik';
import { useModal } from 'components/modal';
import { ConfirmActionModal } from 'astro_2.0/components/ConfirmActionModal';

import { Bounty, BountyContext, BountyProposal } from 'types/bounties';
import { VoteAction } from 'types/proposal';

import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';

import { useWalletContext } from 'context/WalletContext';
import { PaginationResponse } from 'types/api';
import { VOTE_ACTION_SOURCE_PAGE } from 'constants/votingConstants';

export function useBountyControls(
  daoId: string,
  bounty?: Bounty
): {
  handleClaim: () => void;
  handleUnclaim: () => void;
  loading: boolean;
} {
  const router = useRouter();
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();
  const [loading, setLoading] = useState(false);

  const [showModal] = useModal(ConfirmActionModal);

  const onSuccessHandler = useCallback(async () => {
    router.reload();
  }, [router]);

  const handleClaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Claim',
    });

    if (res?.length && bounty) {
      setLoading(true);

      const dao = await SputnikHttpService.getDaoById(daoId);

      if (!dao) {
        return;
      }

      try {
        setVoteActionSource(pathname);

        await nearService?.claimBounty(daoId, {
          bountyId: bounty?.bountyId,
          deadline: bounty?.maxDeadline,
          bountyBond: dao.policy.bountyBond,
          gas: res[0],
          tokenId: bounty?.token,
        });

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    daoId,
    setVoteActionSource,
    pathname,
    nearService,
    onSuccessHandler,
  ]);

  const handleUnclaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Unclaim',
    });

    if (res?.length && bounty) {
      try {
        setLoading(true);
        setVoteActionSource(pathname);
        await nearService?.unclaimBounty(daoId, bounty?.bountyId);

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    setVoteActionSource,
    pathname,
    nearService,
    daoId,
    onSuccessHandler,
  ]);

  return {
    handleClaim,
    handleUnclaim,
    loading,
  };
}

export function useBountyVoting(
  daoId: string,
  proposal: BountyProposal
): {
  handleVote: (vote: VoteAction) => void;
  loading: boolean;
} {
  const router = useRouter();
  const [showModal] = useModal(ConfirmActionModal);
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();

  const [{ loading }, handleVote] = useAsyncFn(
    async (vote: VoteAction) => {
      const res = await showModal({
        title: 'Confirm Your Vote',
      });

      if (res?.length) {
        setVoteActionSource(pathname);
        await nearService?.vote(daoId, proposal.proposalId, vote, res[0]);
        await router.reload();
      }
    },
    [daoId, proposal, router, nearService, pathname, setVoteActionSource]
  );

  return {
    handleVote,
    loading,
  };
}

export function useBountySearch(): {
  handleSearch: (
    val: string
  ) => Promise<PaginationResponse<BountyContext[]> | null>;
  loading: boolean;
} {
  const router = useRouter();
  const daoId = router.query.dao as string;
  const { accountId } = useWalletContext();
  const cancelTokenRef = useRef<CancelTokenSource | null>(null);

  const [{ loading }, handleSearch] = useAsyncFn(async query => {
    if (cancelTokenRef.current) {
      cancelTokenRef.current?.cancel('Cancelled by new req');
    }

    const { CancelToken } = axios;
    const source = CancelToken.source();

    cancelTokenRef.current = source;

    return SputnikHttpService.findBountyContext({
      daoId: daoId || '',
      accountId,
      query,
      cancelToken: source.token,
    });
  }, []);

  return {
    handleSearch,
    loading,
  };
}

export function useHideBounty(): {
  handleSubmit: () => void;
  handleSelect: (id: string) => void;
  selected: string[];
  loading: boolean;
} {
  const router = useRouter();
  const showHidden = router.query?.bountyFilter === 'hidden';
  const daoId = router.query.dao as string;
  const { accountId, nearService, pkAndSignature } = useWalletContext();
  const [selected, setSelected] = useState<string[]>([]);

  const [{ loading }, handleSubmit] = useAsyncFn(async () => {
    if (!pkAndSignature) {
      return;
    }

    try {
      const { publicKey, signature } = pkAndSignature;

      if (publicKey && signature) {
        await SputnikHttpService.toggleBountyContexts({
          accountId,
          publicKey,
          signature,
          daoId,
          ids: selected,
          isArchived: !showHidden,
        });

        await router.replace(router.asPath, undefined, {
          shallow: false,
        });

        setSelected([]);
      }
    } catch (err) {
      showNotification({
        type: NOTIFICATION_TYPES.ERROR,
        lifetime: 20000,
        description: err.message,
      });
    }
  }, [
    selected,
    daoId,
    showHidden,
    accountId,
    router,
    nearService,
    pkAndSignature,
  ]);

  const handleSelect = useCallback(
    (id: string) => {
      if (selected.includes(id)) {
        setSelected(selected.filter(item => item !== id));
      } else {
        setSelected([...selected, id]);
      }
    },
    [selected]
  );

  return {
    handleSubmit,
    handleSelect,
    selected,
    loading,
  };
}
