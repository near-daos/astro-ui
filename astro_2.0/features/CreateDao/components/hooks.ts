import {
  useFormContext,
  UseFormGetValues,
  UseFormSetValue,
} from 'react-hook-form';
import {
  DAOFormValues,
  DAOProposalsType,
  DaoSettingOption,
  DAOStructureType,
  DAOVotingPowerType,
} from 'astro_2.0/features/CreateDao/components/types';
import { useCallback, useEffect, useState } from 'react';
import {
  DAO_PROPOSALS_OPTIONS,
  DAO_STRUCTURE_OPTIONS,
  DAO_VOTING_POWER_OPTIONS,
} from 'astro_2.0/features/CreateDao/components/data';
import { httpService } from 'services/HttpService';
import { CreateDaoCustomInput, CreateDaoInput } from 'types/dao';
import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';
import { configService } from 'services/ConfigService';
import { SputnikWalletError } from 'errors/SputnikWalletError';
import { useRouter } from 'next/router';
import { useTranslation } from 'next-i18next';
import {
  getInitialValues,
  updateAction,
} from 'astro_2.0/features/CreateDao/components/helpers';
import { useStateMachine } from 'little-state-machine';
import { useWalletContext } from 'context/WalletContext';
import { GA_EVENTS, sendGAEvent } from 'utils/ga';

type DaoFormStateReturn = {
  options: DaoSettingOption<
    DAOStructureType | DAOVotingPowerType | DAOProposalsType
  >[];
  getValues: UseFormGetValues<DAOFormValues>;
  setValue: UseFormSetValue<DAOFormValues>;
};

export const useDaoFormState = (): DaoFormStateReturn => {
  const { getValues, watch, setValue } = useFormContext<DAOFormValues>();

  const [structure, voting, proposals] = watch([
    'structure',
    'voting',
    'proposals',
  ]);

  const [options, setOptions] = useState<
    DaoSettingOption<DAOStructureType | DAOVotingPowerType | DAOProposalsType>[]
  >([]);

  useEffect(() => {
    setOptions([
      DAO_STRUCTURE_OPTIONS[structure],
      DAO_VOTING_POWER_OPTIONS[voting],
      DAO_PROPOSALS_OPTIONS[proposals],
    ]);
  }, [structure, voting, proposals, setValue]);

  return { options, getValues, setValue };
};

export function useImageUpload(): {
  uploadImage: (file: File) => Promise<string>;
} {
  const uploadImage = useCallback(async (img: File) => {
    try {
      if (img) {
        const { data } = await httpService.post('/api/upload-to-s3', img, {
          baseURL: '',
        });

        return data;
      }
    } catch (e) {
      return '';
    }

    return '';
  }, []);

  return { uploadImage };
}

export function useCreateDao(): {
  uploadAssets: (defaultFlag: string) => Promise<string>;
  createDao: (
    daoName: string,
    data: CreateDaoInput | CreateDaoCustomInput
  ) => Promise<void>;
} {
  const router = useRouter();
  const { t } = useTranslation();
  const { accountId, nearService } = useWalletContext();
  const { actions, state } = useStateMachine({ updateAction });

  const uploadImg = useCallback(async (img: File) => {
    if (img) {
      const { data } = await httpService.post('/api/upload-to-s3', img, {
        baseURL: '',
      });

      return data;
    }

    return '';
  }, []);

  async function loadImage(defaultFlag: string) {
    const response = await fetch(defaultFlag);
    const blob = await response.blob();

    return new File([blob], 'image.png', { type: 'image/png' });
  }

  const uploadAssets = useCallback(
    async (defaultFlag: string) => {
      try {
        const defaultFlagFile = await loadImage(defaultFlag);

        return uploadImg(defaultFlagFile);
      } catch (e) {
        return '';
      }
    },
    [uploadImg]
  );

  const createDao = useCallback(
    async (daoName: string, data: CreateDaoInput | CreateDaoCustomInput) => {
      try {
        await nearService?.createDao(data);

        const { nearConfig } = configService.get();
        const daoId = `${daoName}.${nearConfig?.contractName ?? ''}`;

        sendGAEvent({
          name: GA_EVENTS.CREATE_DAO,
          daoId,
          accountId,
        });

        showNotification({
          type: NOTIFICATION_TYPES.INFO,
          description: t('notifications.transactionDelay'),
          lifetime: 20000,
        });

        sessionStorage.removeItem('__LSM__');
        actions.updateAction(
          getInitialValues(accountId, state.assets.defaultFlag)
        );

        await router.push(`/dao/${daoId}`);
      } catch (error) {
        console.warn(error);

        if (error instanceof SputnikWalletError) {
          showNotification({
            type: NOTIFICATION_TYPES.ERROR,
            description: error.message,
            lifetime: 20000,
          });
        }
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [router, uploadImg, nearService]
  );

  return {
    createDao,
    uploadAssets,
  };
}
