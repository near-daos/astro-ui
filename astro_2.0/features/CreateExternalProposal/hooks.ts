import { ProposalVariant } from 'types/proposal';
import { useCallback, useEffect, useState } from 'react';
import { useWalletContext } from 'context/WalletContext';
import { useMountedState } from 'react-use';
import { useRouter } from 'next/router';
import omit from 'lodash/omit';

const CREATE_PROPOSAL = 'create_proposal';

// example:
// https://dev.app.astrodao.com/dao/aviarium.sputnikv2.testnet/proposals?action=create_proposal&variant=ProposeCustomFunctionCall&params={"details":"Proposal description here", "methodName": "create", "smartContractAddress": "sputnikv2.testnet", "json":{ "key": "some value" }}

export function isActionValid(action: string | undefined): boolean {
  return action === CREATE_PROPOSAL;
}

export function isVariantValid(proposalVariant: string | undefined): boolean {
  return Object.values(ProposalVariant).includes(
    proposalVariant as ProposalVariant
  );
}

export function useCreateProposalFromExternal(
  onCreateProposal: (
    initialProposalVariant?: ProposalVariant,
    initialValues?: Record<string, unknown>
  ) => void
): {
  error: string;
  handleDismiss: () => void;
} {
  const router = useRouter();
  const isMounted = useMountedState();
  const { accountId } = useWalletContext();
  const { action, variant, params } = router.query;

  const [error, setError] = useState<string>('');

  const handleDismiss = useCallback(() => {
    setError('');
    router.replace(
      {
        pathname: router.pathname,
        query: omit(router.query, ['action', 'variant', 'params']),
      },
      undefined,
      {
        shallow: true,
      }
    );
  }, [router]);

  useEffect(() => {
    (async () => {
      try {
        if (!action && !variant) {
          return;
        }

        if (!isActionValid(action as string)) {
          setError('Unsupported action');

          return;
        }

        if (!isVariantValid(variant as string)) {
          setError('Unsupported proposal variant');

          return;
        }

        if (!accountId) {
          setError(
            'Cannot create proposal by unauthorized user. Please sign in first'
          );

          return;
        }

        // http://localhost:8080/dao/aviarium.sputnikv2.testnet/proposals?action=create_proposal&variant=ProposeCustomFunctionCall&params={"details":"Proposal description here", "methodName": "create", "smartContractAddress": "sputnikv2.testnet", "json":{ "msg": "{\"draft_group_id\":1}" }}
        // http://localhost:8080/dao/aviarium.sputnikv2.testnet/proposals?action=create_proposal&variant=ProposeCustomFunctionCall&params={"details":"Proposal description here", "methodName": "create", "smartContractAddress": "sputnikv2.testnet", "json":{ "msg": {"draft_group_id":1} }}

        const parsedParams = params
          ? JSON.parse(decodeURIComponent(params as string).trim())
          : {};

        setError('');

        setTimeout(() => {
          if (!isMounted()) {
            return;
          }

          onCreateProposal(variant as ProposalVariant, {
            ...parsedParams,
            json: JSON.stringify(parsedParams.json),
          });
        }, 1500);
      } catch (e) {
        // probably invalid params
        setError('Invalid proposal parameters');
      }
    })();
  }, [onCreateProposal, accountId, isMounted, action, variant, params]);

  return {
    error,
    handleDismiss,
  };
}
