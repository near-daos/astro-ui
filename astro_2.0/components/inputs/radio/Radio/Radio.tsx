import React from 'react';
import cn from 'classnames';

import { useRadioContext } from './hooks';

import styles from './Radio.module.scss';

export type RadioProps = {
  className?: string;
  id?: string;
  value: string;
  label: React.ReactNode;
  disabled?: boolean;
  type?: string;
};

export const Radio: React.FC<RadioProps> = ({
  className,
  value,
  label,
  disabled = false,
  id,
  type,
}: RadioProps) => {
  const {
    itemClassName,
    activeItemClassName = '',
    state,
    onChange,
  } = useRadioContext();

  const checked = value === state;
  const idExt = id || `input-radio-${value}`;

  return (
    <label
      htmlFor={idExt}
      className={cn(styles.root, className, itemClassName, {
        [styles.disabled]: disabled,
        [activeItemClassName]: checked,
        [styles.notifications]: type === 'notifications',
      })}
    >
      <input
        id={idExt}
        value={value}
        checked={checked}
        type="radio"
        className={cn(styles.input, { [styles.checked]: checked })}
        disabled={disabled}
        onChange={e => onChange(e.target.value, e)}
      />
      <span className={styles.checkmark} />
      <span className={cn(styles.label)}>{label}</span>
    </label>
  );
};
