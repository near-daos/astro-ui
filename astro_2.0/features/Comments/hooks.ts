import { useCallback, useEffect, useState } from 'react';
import { useRouter } from 'next/router';
import uniqBy from 'lodash/uniqBy';
import { useAsyncFn, useMount, useMountedState } from 'react-use';
import { useDraftsContext } from 'astro_2.0/features/Drafts/components/DraftsProvider';
import { DraftComment } from 'services/DraftsService/types';
import { useWalletContext } from 'context/WalletContext';
import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';
import { configService } from 'services/ConfigService';
import io, { Socket as TSocket } from 'socket.io-client';
import { fetcher as getDraftProposalComments } from 'services/ApiService/hooks/useDraftComments';
import { useFlags } from 'launchdarkly-react-client-sdk';
import { DraftCommentIndex } from 'services/SearchService/types';
import { mapDraftCommentIndexToDraftComment } from 'services/SearchService/mappers/draftComment';

type Socket = typeof TSocket;

function prepareData(data: DraftComment[]) {
  const groups = data.reduce<Record<string, DraftComment[]>>((res, item) => {
    const key = item.replyTo ?? '';

    if (res[key]) {
      res[key].push(item);
    } else {
      res[key] = [item];
    }

    return res;
  }, {});

  if (!groups['']) {
    return data;
  }

  return groups[''].map(item => {
    return {
      ...item,
      replies: groups[item.id] ?? [],
    };
  });
}

export function useDraftComments(): {
  loading: boolean;
  countComments: number;
  data: DraftComment[];
  addComment: (val: string) => Promise<void>;
  editComment: (val: string, id: string) => Promise<void>;
  deleteComment: (id: string) => Promise<void>;
  likeComment: (id: string, isLiked: boolean) => Promise<void>;
  dislikeComment: (id: string, idDislike: boolean) => Promise<void>;
} {
  const isMounted = useMountedState();
  const router = useRouter();
  const { draft, dao } = router.query;
  const contextId = draft as string;
  const daoId = dao as string;
  const contextType = 'DraftProposal';

  const { useOpenSearchDataApi } = useFlags();
  const { draftsService, setAmountComments } = useDraftsContext();
  const { accountId, pkAndSignature } = useWalletContext();

  const [value, setValue] = useState<{
    data: DraftComment[];
    countComments: number;
  }>({
    data: [],
    countComments: 0,
  });

  const [{ loading }, getAllComments] = useAsyncFn(async () => {
    if (useOpenSearchDataApi === undefined) {
      return;
    }

    try {
      const data = useOpenSearchDataApi
        ? await getDraftProposalComments(
            'draftProposalComment',
            daoId,
            contextId
          )
        : await draftsService.getDraftComments({
            contextId,
            contextType,
            offset: 0,
            limit: 1000,
          });

      setValue({
        data,
        countComments: data.length,
      });
    } catch (e) {
      showNotification({
        type: NOTIFICATION_TYPES.ERROR,
        lifetime: 20000,
        description: e?.message,
      });
    }
  }, [contextId, useOpenSearchDataApi]);

  useMount(() => getAllComments());

  const addComment = useCallback(
    async (msg: string, replyTo?: string) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      try {
        await draftsService.createDraftComment({
          contextId,
          daoId,
          contextType,
          message: msg,
          replyTo,
          accountId,
          publicKey,
          signature,
        });
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [accountId, contextId, daoId, draftsService, pkAndSignature]
  );

  const editComment = useCallback(
    async (msg: string, id: string) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      try {
        await draftsService.editDraftComment({
          id,
          daoId,
          draftId: contextId,
          message: msg,
          accountId,
          publicKey,
          signature,
        });

        await getAllComments();
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [pkAndSignature, draftsService, daoId, contextId, accountId, getAllComments]
  );

  const likeComment = useCallback(
    async (id: string, isLiked: boolean) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      const params = {
        id,
        daoId,
        draftId: contextId,
        accountId,
        publicKey,
        signature,
      };

      try {
        if (isLiked) {
          await draftsService.removeLikeDraftComment(params);
        } else {
          await draftsService.removeDislikeDraftComment(params);
          await draftsService.likeDraftComment(params);
        }
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [accountId, contextId, daoId, draftsService, pkAndSignature]
  );

  const dislikeComment = useCallback(
    async (id: string, isDislike: boolean) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      const params = {
        id,
        daoId,
        draftId: contextId,
        accountId,
        publicKey,
        signature,
      };

      try {
        if (isDislike) {
          await draftsService.removeDislikeDraftComment(params);
        } else {
          await draftsService.removeLikeDraftComment(params);
          await draftsService.dislikeDraftComment(params);
        }
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [accountId, contextId, daoId, draftsService, pkAndSignature]
  );

  const deleteComment = useCallback(
    async (id: string) => {
      const { publicKey, signature } = pkAndSignature || {};

      if (!publicKey || !signature) {
        return;
      }

      try {
        await draftsService.deleteDraftComment({
          id,
          daoId,
          draftId: contextId,
          accountId,
          publicKey,
          signature,
        });
      } catch (e) {
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 20000,
          description: e?.message,
        });
      }
    },
    [accountId, contextId, daoId, draftsService, pkAndSignature]
  );

  useEffect(() => {
    setAmountComments(value?.countComments || 0);
  }, [value?.countComments, setAmountComments]);

  useEffect(() => {
    let socket: Socket;
    const { appConfig } = configService.get();

    const { publicKey, signature } = pkAndSignature || {};

    if (accountId && publicKey && isMounted() && appConfig) {
      socket = io(appConfig.DRAFTS_API_URL, {
        query: {
          accountId,
          publicKey,
          signature,
        },
        transports: ['websocket'],
      });

      if (socket) {
        socket.on(
          'draft-comment',
          (comment: DraftComment | DraftCommentIndex) => {
            if (isMounted()) {
              setValue(prev => {
                const newValue = mapDraftCommentIndexToDraftComment(
                  comment as DraftCommentIndex
                ) as DraftComment;

                return {
                  data: uniqBy([newValue, ...prev.data], item => item.id),
                  countComments: prev.countComments + 1,
                };
              });
            }
          }
        );
        socket.on(
          'draft-comment-updated',
          (comment: DraftComment | DraftCommentIndex) => {
            if (isMounted()) {
              setValue(prev => {
                const newValue = mapDraftCommentIndexToDraftComment(
                  comment as DraftCommentIndex
                ) as DraftComment;

                const newData = prev.data.map(item => {
                  if (item.id !== newValue.id) {
                    return item;
                  }

                  return newValue;
                });

                return {
                  data: newData,
                  countComments: newData.length,
                };
              });
            }
          }
        );
        socket.on(
          'draft-comment-removed',
          (comment: DraftComment | DraftCommentIndex) => {
            if (isMounted()) {
              const newValue = mapDraftCommentIndexToDraftComment(
                comment as DraftCommentIndex
              ) as DraftComment;

              setValue(prev => {
                const newData = prev.data.filter(
                  item => item.id !== newValue.id
                );

                return {
                  data: newData,
                  countComments: newData.length,
                };
              });
            }
          }
        );
      }
    }
  }, [accountId, isMounted, pkAndSignature, useOpenSearchDataApi]);

  const preparedData = prepareData(value.data);

  return {
    loading,
    countComments: preparedData?.length || 0,
    data: preparedData ?? [],
    addComment,
    editComment,
    deleteComment,
    likeComment,
    dislikeComment,
  };
}
