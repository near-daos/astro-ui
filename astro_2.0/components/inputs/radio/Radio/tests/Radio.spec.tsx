/* eslint-disable @typescript-eslint/ban-ts-comment */

import { render } from 'jest/testUtils';
import { fireEvent } from '@testing-library/dom';

import { Radio } from 'astro_2.0/components/inputs/radio/Radio';

import { useRadioContext } from 'astro_2.0/components/inputs/radio/Radio/hooks';

jest.mock('astro_2.0/components/inputs/radio/Radio/hooks', () => {
  return {
    useRadioContext: jest.fn(() => ({})),
  };
});

describe('Radio', () => {
  it('Should render component', () => {
    const { container } = render(<Radio value="Hello" label="World" />);

    expect(container).toMatchSnapshot();
  });

  it('Should call onChange', () => {
    const value = 'Hello';

    const onChange = jest.fn();

    // @ts-ignore
    useRadioContext.mockImplementation(() => ({
      onChange,
    }));

    const { getByRole } = render(<Radio value={value} label="World" />);

    fireEvent.click(getByRole('radio'));

    expect(onChange).toBeCalledWith(value, expect.anything());
  });
});
