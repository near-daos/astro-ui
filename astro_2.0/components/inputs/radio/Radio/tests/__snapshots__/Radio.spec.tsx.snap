// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Should render component 1`] = `
<div>
  <label
    class="root"
    for="input-radio-Hello"
  >
    <input
      class="input"
      id="input-radio-Hello"
      type="radio"
      value="Hello"
    />
    <span
      class="checkmark"
    />
    <span
      class="label"
    >
      World
    </span>
  </label>
</div>
`;
